business/istio_validations.go

```diff
@@ -15,135 +15,168 @@ import (
-	defer timer.ObserveDuration()
+	return models.IstioValidations(validationsForCluster(in.kialiCache.Validations().Items(), cluster)).FilterBySingleType(schema.GroupVersionKind{Group: "", Version: "", Kind: "service"}, service), nil
+}
 
-	wg := sync.WaitGroup{}
-	errChan := make(chan error, 1)
+func (in *IstioValidationsService) GetValidationsForWorkload(ctx context.Context, cluster, namespace, workload string) (models.IstioValidations, error) {
+	// Check if user has access to the namespace (RBAC) in cache scenarios and/or
+	// if namespace is accessible from Kiali (Deployment.AccessibleNamespaces)
+	if namespace != "" {
+		if _, err := in.namespace.GetClusterNamespace(ctx, namespace, cluster); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("Namespace param should be set for Validations in cluster %s", cluster)
+	}
+
+	return models.IstioValidations(validationsForCluster(in.kialiCache.Validations().Items(), cluster)).FilterBySingleType(schema.GroupVersionKind{Group: "", Version: "", Kind: "workload"}, workload), nil
+}
+
+// CreateValidations returns an IstioValidations object with all the checks found when running
+// all the enabled checkers. If service is "" then the whole namespace is validated.
+// If service is not empty string, then all of its associated Istio objects are validated.
+func (in *IstioValidationsService) CreateValidations(ctx context.Context, cluster string) (models.IstioValidations, error) {
+	var end observability.EndFunc
+	ctx, end = observability.StartSpan(ctx, "getValidations",
+		observability.Attribute("package", "business"),
+		observability.Attribute("cluster", cluster),
+	)
+	defer end()
+
+	timer := internalmetrics.GetValidationProcessingTimePrometheusTimer("", "")
+	defer timer.ObserveDuration()
 
-	var istioConfigList models.IstioConfigList
-	var services models.ServiceList
 	var serviceAccounts map[string][]string
 	var namespaces models.Namespaces
-	var workloadsPerNamespace map[string]models.WorkloadList
-	var mtlsDetails kubernetes.MTLSDetails
-	var rbacDetails kubernetes.RBACDetails
 	var registryServices []*kubernetes.RegistryService
+	var workloadsPerNamespace map[string]models.WorkloadList
 
-	wg.Add(3) // We need to add these here to make sure we don't execute wg.Wait() before scheduler has started goroutines
-	if service != "" {
-		wg.Add(1)
-	}
-
-	// We fetch without target service as some validations will require full-namespace details
-	go in.fetchIstioConfigList(ctx, &istioConfigList, &mtlsDetails, &rbacDetails, cluster, namespace, errChan, &wg)
-
-	if workload != "" {
-		// load only requested workload
-		go in.fetchWorkload(ctx, &workloadsPerNamespace, cluster, workload, namespace, errChan, &wg)
-	} else {
-		go in.fetchAllWorkloads(ctx, &workloadsPerNamespace, cluster, &namespaces, errChan, &wg)
+	err := in.fetchAllWorkloadsDirect(ctx, &workloadsPerNamespace, cluster, &namespaces)
```

Comment:
Do we need two different `fetchAllWorkloads` methods?

In the other PR, it uses `GetWorkloadList`: https://github.com/kiali/kiali/pull/7136/files#diff-4afe91ed13da4c9e145fdccf0596bee0b6c0fc2e29d3da22f112584428159dd3R159. Is there a difference between the two?

Reply:
It looks like inside of `fetchAllWorkloadsDirect` we set the value of `&workloadsPerNamespace` and `&namespaces` but we also need `namespaces` to be set further down. Why not just set `namepaces` in this method and pass those down into `fetchAllWorkloads(ctx, cluster, namespaces)`? `fetchAllWorkloads` can return the `map[string]models.WorkloadList` and then you can get around needing to pass a pointer to a map into those methods which is a little awkward to grok.

Reply:
`fetchAllWorkloads` is a wrapper method for `fetchAllWorkloadsDirect` passing all params and designed for async call, while `fetchAllWorkloadsDirect` is for sync call.
For `GetIstioObjectValidations` we use async calls of fetch methods, and I kept those async calls order and method params for 'CreateValidations' sync calls to be consistent and easy to read the code. For this have created 'Direct' methods for sync calls, and the existing methods are now still async but calling newly created direct sync versions of methods and pass all params + WaitGroup, error.

Reply:
I think it's ok that `GetIstioObjectValidations` and `CreateValidations` are different. Having this wrapper method makes things harder to follow. If we want both methods to be the same then we can update both rather than having the additional wrapper on top. I don't remember why I updated `GetValidations` which turned into `CreateValidations` in the first place. Probably async fetching of different things is not necessary since everything is in memory already.

Reply:
Yes async fetching objects for `CreateValidations` and cache was causing some backend errors after some time, so the sync call is necessary there.
However  the `GetIstioObjectValidations` still needs async call for performance reasons.

Reply:
> Yes async fetching objects for CreateValidations and cache was causing some backend errors after some time, so the sync call is necessary there.
However the GetIstioObjectValidations still needs async call for performance reasons.

Can we keep the two separate then without the wrappers?

Reply:
There is one new function `fetchAllWorkloadsDirect` which does the main job and is called sync only, it has the core of the original function `fetchAllWorkloads`. And `fetchAllWorkloads` now is a wrapper for async call for  `fetchAllWorkloadsDirect`.

Reply:
Can we keep `fetchAllWorkloads` the same as it was and remove the `fetchAll...Direct` methods? If we need to keep `CreateValidations` sync, we don't need to call `fetchAllWorkloads` at all in `CreateValidations` and can just put that logic in directly like how it is here: https://github.com/kiali/kiali/pull/7136/files#diff-4afe91ed13da4c9e145fdccf0596bee0b6c0fc2e29d3da22f112584428159dd3R151-R164. Maybe there's another way of organizing things that also is clearer.

The two things that are hard for me to follow are 1. passing the return values _into_ the function as opposed to just returning them _from_ the function. This also means you need to pass pointers to maps/slices etc. which generally isn't needed and also if you change the order of the function calls then it will affect the return value args but you'll get no compiler error. 2. having multiple wrappers for functions. This makes it hard for me to follow where things are being set.

I actually think if you pull all the async bits out of `fetchAllWorkloads` then you can use it in both `CreateValidations` and `GetIstioObjectValidations` without needing the wrappers. Something like this:
```
go func() {
		defer wg.Done()

		if len(errChan) > 0 {
			return
		}
		var fetchErr error
		if istioConfigList, mtlsDetails, rbacDetails, fetchErr = in.fetchIstioConfigList(ctx, cluster, namespace); err != nil {
			errChan<-err
		} 
	}()
```
Then you can call `fetchIstioConfigList` the same way in `CreateValidations`

```
	for _, namespace := range namespaces {
		istioConfigs, mtlsDetails, rbacDetails, err := in.fetchIstioConfigList(ctx, cluster, namespace.Name)
		if err != nil {
			return nil, err
		}
        ...
```

Reply:
I see now, thanks for the suggestion, have removed the wrapper methods and now only one method is left for direct calls. However the reference variables seems to be the part of sync calls logic in `GetIstioObjectValidations` methods.